### DISCUSSION TITLE
Configure Lsp and Git getters

### DISCUSSION THREAD

**Discussion by man9ourah** [nvim-neo-tree/neo-tree.nvim#127] [Q&A]
Can the user provide functions to return the lsp and git information? i.e. I want to override utils.get_diagnostic_counts to use CoC. It would be super helpful if the setup function can take functions to obtain the git and Lsp info .. complying with whatever api currently used by builtin functions.

**Comment by cseickel** [comment]
Yes, you can do that. The way it works now is that it reacts to nvim's DiagnosticChanged event. Is there a similar event for CoC?

The general idea is to call:
```lua
require("neo-tree.sources.manager").diagnostics_changed("filesystem", { 
  diagnostics_lookup = {
    "path/to/file" = { severity_number = 1, severity_string = "Error" },
    "path/to/file2" = { severity_number = 2, severity_string = "Warning" }
  }
}}
```
... whenever the diagnostic counts change. Here is the built-in function that creates the lookup from LSP Disagnostics:
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/caa64cf0e43cf141bc6d34ff6216c1ab2611d75c/lua/neo-tree/utils.lua#L104

**Comment by man9ourah** [comment]
Thanks for your response. 

Yes, Coc has ‘User CocDiagnosticChange’. So I should call ‘ require("neo-tree.sources.manager").diagnostics_changed’ with a table of diagnostics on that event, right? 

How about the git information?

**Comment by cseickel** [comment]
Actually, there might be a cleaner way to replace the built-in diagnostics. You can replace the built in event handler AFTER you call the setup() function like this:

```lua
  require("neo-tree").setup({ 
    enable_diagnostics = true
    ... 
  })

  local events = require("neo-tree.events")
  events.destroy_event(events.VIM_DIAGNOSTIC_CHANGED)
  events.define_autocmd_event(events.VIM_DIAGNOSTIC_CHANGED, { "User CocDiagnosticChange" }, 500, function(args)
    args.diagnostics_lookup = utils.get_diagnostic_counts() -- <-- your handler here
    return args
  end)
```
Doing it that way, you won't need to call `diagnostics_changed` manually.


For the git status data, there is no git event so I just pull that data before every render using the neo-tree `"before_render"` event. Here is how to replace that:

```lua
      require("neo-tree").setup({
        enable_git_status = false,
        event_handlers = {
          {
            event = "before_render",
            handler = function(state)
              state.git_status_lookup = require("neo-tree.git").status() -- <-- replace with your handler
            end
          },
        }
      })
```

The built-in git status provider is at: https://github.com/nvim-neo-tree/neo-tree.nvim/blob/v1.x/lua/neo-tree/git/init.lua

**Comment by man9ourah** [comment]
Thank you so much for the help. Indeed this looks like a cleaner way to override those functionalities. Also, thanks for this plugin, I appreciate the extra efforts making it so customizable. Closing this issue.

**Comment by cseickel** [comment]
You're very welcome! One other thing you may want to consider is that you don't have to use the built-in components to display this either. if you want to build the lookups differently or display different info, you can create a custum component. the wiki has some examples:
https://github.com/nvim-neo-tree/neo-tree.nvim/wiki/Recipes#components

and of course: 
```
:h neo-tree-components
```

Let me know if you run into any issues I can help with. If you come up with a good solution, please consider posting it back here or the Wiki. I am converting this to a discussion for others to find.

**Reply by radoslawroszkowiak** [reply]
Hi @cseickel 
Is this ` events.define_autocmd_event(events.VIM_DIAGNOSTIC_CHANGED...)` advice still valid with the current neo-tree version?
I'm trying to follow that to set up CoC diagnostics, but it looks like `events.destroy_event(events.VIM_DIAGNOSTIC_CHANGED)` doesn't really destroy that. After executing:
```
local events = require("neo-tree.events")
events.destroy_event(events.VIM_DIAGNOSTIC_CHANGED)
events.define_autocmd_event(events.VIM_DIAGNOSTIC_CHANGED, {
```
I'm getting the following:
```
E5113: Error while calling lua chunk: ...packer/start/neo-tree.nvim/lua/neo-tree/events/queue.lua:151: Event already defined: vim_diagnostic_changed
stack traceback:
        [C]: in function 'error'
        ...packer/start/neo-tree.nvim/lua/neo-tree/events/queue.lua:151: in function 'define_event'
        .../packer/start/neo-tree.nvim/lua/neo-tree/events/init.lua:77: in function 'define_autocmd_event'
        ...e/rroszkowiak/.config/nvim/after/plugin/nvim-tree.rc.lua:587: in main chunk

```

BTW, do you think there is a place for integrating CoC diagnostics with your plugin? I.e. nvim-tree currently lets the user choose between LSP and CoC diagnostics provider via a simple config switch.
If so, I could possibly try to implement it (even though my lua and nvim plugins experience is basically none).

And thanks for the great plugin, it makes the best impression of the tree plugins I tried so far!

**Reply by cseickel** [reply]
That code should work, it seems like there is a bug somewhere.

I think it should be relatively simple to add CoC diagnostics, so if you wanted to do it I would accept the PR so long as it is as simple as I expect. I certainly did design the system with that option in mind.
